osrf/mbzirc
# Extend GameLogicPlugin to support scoring manipulation task

New logic added in GameLogicPlugin:
* checks for if any target objects are placed inside the specified regions over the USV
    *  Done using a new `EntityDetector` plugin - code is adapted from [PerfromerDetector](https://github.com/ignitionrobotics/ign-gazebo/tree/ign-gazebo6/src/systems/performer_detector)  but works for all entities.
* user gets 2 tries (hence there are now 2 small and 2 large objects on the target vessel). They only need to place one of each on the USV.
*  objects retrieval must be done in the correct order: small object -> large object
* if object drops into ocean, user gets a time penalty

The target object region placement region is currently quite generous. Green boxes below visualizes the valid region of placement:

![mbzirc_object_placement_region](https://user-images.githubusercontent.com/4000684/161661553-22637a1e-caef-4ff6-8105-900b1292c08c.png)


To test:

1. launch coast world:

    ```
    ros2 launch mbzirc_ros competition.launch.py ign_args:="-v 4 -r coast.sdf"
    ```

1. spawn the USV

    ```
    ros2 launch mbzirc_ign spawn.launch.py name:=usv world:=coast model:=usv type:=usv x:=-1462 y:=-16.5 z:=0.3 R:=0 P:=0 Y:=0
    ```

1. skip the target identification and go straight to the end of identification phase:

    ```
    ign service -s /mbzirc/skip_to_phase --reqtype ignition.msgs.StringMsg --reptype ignition.msgs.Boolean --timeout 300 --req 'data: "large_object_id_success"'
    ```

    You can confirm this by checking the current phase using `ros2 topic echo /mbzirc/phase`

1. Open separate terminals and monitor `events.yml` and `score.yml` logs

    ```
    tail -f  /tmp/ign/mbzirc/logs/events.yml
    ```

    ```
    watch -n 1 -d cat /tmp/ign/mbzirc/logs/score.yml
    ```

1. Test unsuccessful small object retrieval. Drop `small_box_a` into ocean:

    ```
    ign service -s /world/coast/set_pose --reqtype ignition.msgs.Pose --reptype ignition.msgs.Boolean --timeout 300 --req 'name: "small_box_a", position: {x:-1462, y:-10, z: 2.0}'
    ```

    Wait a few secs and you should see a `small_object_retrieve_failure_1` event logged in `events.yml` and the score in `score.yml` jumps by 120s

1. Test successful small object retrieval. Drop `small_box_a_duplicate` onto USV:

    ```
    ign service -s /world/coast/set_pose --reqtype ignition.msgs.Pose --reptype ignition.msgs.Boolean --timeout 300 --req 'name: "small_box_a_duplicate", position: {x:-1462, y:-17.85, z: 2.0}'
    ```

    You should see a `small_object_retrieve_success` event logged in `events.yml` and the score continues to increment by 1s at a time.


1. Test unsuccessful large object retrieval. Drop `large_dry_box_a` into ocean:

    ```
    ign service -s /world/coast/set_pose --reqtype ignition.msgs.Pose --reptype ignition.msgs.Boolean --timeout 300 --req 'name: "large_dry_box_a", position: {x:-1461, y:-10, z: 2.0}'
    ```

    Wait a few secs and you should see a `large_object_retrieve_failure_1` event logged in `events.yml` and the score in `score.yml` jumps by 120s

1. Test successful large object retrieval. Drop `large_dry_box_a_duplicate` onto USV:

    ```
    ign service -s /world/coast/set_pose --reqtype ignition.msgs.Pose --reptype ignition.msgs.Boolean --timeout 300 --req 'name: "large_dry_box_a_duplicate", position: {x:-1462, y:-16.5, z: 2.0}'
    ```

    You should see a `large_object_retrieve_success` event and a `finished` event logged in `events.yml`. Simulation should now be paused to indicate the end of the run.
